src/toxcore/tcp/server/client.rs

```diff
@@ -0,0 +1,215 @@
+use std::slice::Iter;
+
+use futures::{Sink, Future};
+use futures::sync::mpsc;
+
+use tokio_io::IoFuture;
+
+/** Structure that represents how Server keeps connected clients. A write-only socket with
+human interface. A client cannot send a message directly to another client, whereas server can.
+*/
+#[derive(Clone)]
+pub struct Client {
+    /// PublicKey of the client.
+    pk: PublicKey,
+    /// The transmission end of a channel which is used to send values.
+    tx: mpsc::Sender<Packet>,
+    /** links - a table of indexing links from this client to another
+
+    A client requests to link him with another client by PK with RouteRequest.
+    The server inserts that PK into links and gives the index of the link back to client
+    via RouteResponse. Now the client may use this index to communicate with the connection
+    using that index, e.g. send Data by index. Our links are 0-based while wire indices are
+    16-based. E.g. `::get_connection_id` and `::insert_connection_id` return `Some(x+16)`,
+    `::get_link` and `::take_link` accept ids in `[0; 240) + 16`. All conversions are done only
+    inside this module.
+    */
+    links: [Option<PublicKey>; 240],
+    /// Used to check whether PongResponse is correct
+    ping_id: u64
+}
+
+impl Client {
+    /** Create new Client
+    */
+    pub fn new(tx: mpsc::Sender<Packet>, pk: &PublicKey) -> Client {
+        Client {
+            pk: *pk,
+            tx: tx,
+            links: [None; 240],
+            ping_id: 0
+        }
+    }
+
+    /** PK of the `Client`
+    */
+    pub fn pk(&self) -> PublicKey {
+        self.pk
+    }
+
+    /** Last ping_id sent to client.
+    */
+    pub fn ping_id(&self) -> u64 {
+        self.ping_id
+    }
+
+    /** Return index of of the link by PK
+
+    Some(index + 16) if link exists
+
+    None if there is no such PK linked to this client
+    */
+    pub fn get_connection_id(&self, to: &PublicKey) -> Option<u8> {
+        self.links.iter().position(|&link| link == Some(*to)).map(|x| x as u8).map(|x| x + 16)
+    }
+
+    /** Try to link PK.
+
+    Some(index + 16) if has been inserted or link existed
+
+    None if no free space to insert
+    */
+    pub fn insert_connection_id(&mut self, to: &PublicKey) -> Option<u8> {
+        match self.get_connection_id(to) {
+            Some(index) => Some(index), // already inserted
+            None => {
+                if let Some(index) = self.links.iter().position(|link| link.is_none()) {
+                    self.links[index] = Some(*to);
+                    Some(index as u8).map(|x| x + 16)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /** Get link by connection_id.
+    Ensure connection_id [0; 240) + 16
+    */
+    pub fn get_link(&self, connection_id: u8) -> Option<PublicKey> {
+        self.links[connection_id as usize - 16]
+    }
+
+    /** Get link by connection_id and remove it from contaiter.
```

Comment:
It is very trivial thing.
The comments is wrong spelling.
contaiter ==> container